File: csharp/0094-binary-tree-inorder-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> InorderTraversal(TreeNode root) {
        // return InorderTraversalRecursive(root).ToList();
        return InorderTraversalIterative(root).ToList();
    }

    // Time: O(n)
    // Space: O(n)
    private IEnumerable<int> InorderTraversalRecursive(TreeNode node) {
        if (node == null) {
            return Enumerable.Empty<int>();
        }

        return InorderTraversalRecursive(node.left)
            .Append(node.val)
            .Concat(InorderTraversalRecursive(node.right));
    }

    // Time: O(n)
    // Space: O(n)
    private IEnumerable<int> InorderTraversalIterative(TreeNode node) {
        List<int> result = new();
        Stack<TreeNode> stack = new();
        while (node != null || stack.Any()) {
            while (node != null) {
                stack.Push(node);
                node = node.left;
            }
            node = stack.Pop();
            result.Add(node.val);
            node = node.right;
        }
        return result;
    }
}

File: go/0094-binary-tree-inorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
	res := []int{}
	stack := []*TreeNode{}
	cur := root

	for cur != nil || len(stack) > 0 {
		if cur != nil {
			stack = append(stack, cur)
			cur = cur.Left
		} else {
			cur, stack = stack[len(stack)-1], stack[:len(stack)-1]
			res = append(res, cur.Val)
			cur = cur.Right
		}
	}
	return res
}
